Assets/Tests/InputSystem/CorePerformanceTests.cs

```diff
@@ -766,6 +766,66 @@ void CallUpdate()
         }
     }
 
+    [Test, Performance]
+    [Category("Performance")]
+    [TestCase(OptimizationTestType.NoOptimization)]
+    [TestCase(OptimizationTestType.ReadValueCaching)]
+    // These tests show the performance of the ReadValueCaching optimization when there are state changes to gamepad
+    // controls.
+    // Two cases of control changes are evaluated to show the current performance impact:
+    // 1) When the control value changed is a float. In this case there's a performance loss when using
+    // ReadValueCaching.
+    // 2) When the control value changed is a composite control. In this case there's a performance improvement when
+    // using ReadValueCaching.
+    public void Performance_OptimizedControls_EvaluateStaleControlReadsWhenGamepadStateChanges(OptimizationTestType testType)
+    {
+        SetInternalFeatureFlagsFromTestType(testType);
+
+        var gamepad = InputSystem.AddDevice<Gamepad>();
+
+        var value = 0f;
+        var value2d = Vector2.zero;
+        Measure.Method(() =>
+        {
+            for (var i = 0; i < 1000; ++i)
+            {
+                // Make sure state changes are different from previous state so that we mark the controls as
+                // stale.
+                InputSystem.QueueStateEvent(gamepad, new GamepadState { leftTrigger = i / 1000f, rightTrigger = i / 1000f});
```

Comment:
Wouldn't it make sense to not include the queue state event part in the benchmark and instead put this on a background thread? (applies to all the benchmarks). If you do, you can spawn that thread and then let it busy-wait on an atomic until measurement thread signals it and then let them execute. Otherwise this might be dominated by the part that is not subject for testing?

Reply:
In theory that makes sense and it's maybe something we should do. But I think it should be out of scope of this PR as it will be more complex. 
Also, I'm not sure the impact of `InputSystem.QueueStateEvent` matters that much here. It will be the same regardless of the optimization type. Also, the goal here is to do a relative benchmark rather than a absolute one. Even if there's something "coloring" the test with some extra work, it colors all the tests in the same way.